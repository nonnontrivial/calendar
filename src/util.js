// @flow

let dayNames = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
let monthNames = [
  'Jan',
  'Feb',
  'Mar',
  'Apr',
  'May',
  'Jun',
  'July',
  'Aug',
  'Sep',
  'Oct',
  'Nov',
  'Dec'
];

export { dayNames, monthNames };
