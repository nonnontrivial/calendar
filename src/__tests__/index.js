import { render, cleanup, fireEvent } from 'react-testing-library';
import { withProfiler } from 'jest-react-profiler';
import React from 'react';
import { dayNames, monthNames } from '../util';
import { Calendar, MonthCycler, DateContext } from '../';

afterEach(cleanup);

function renderWithContext(Component, value, componentProps = {}) {
  return render(
    <DateContext.Provider value={{ date: new Date(), onChange: jest.fn(), ...value }}>
      <Component {...componentProps} />
    </DateContext.Provider>
  );
}

test('MonthCycler does not overcommit', () => {
  let { container, getByText } = renderWithContext(withProfiler(MonthCycler));
  expect(container.querySelector('p')).toHaveCommittedTimes(undefined);
  fireEvent.click(getByText('←'));
  expect(container.querySelector('p')).toHaveCommittedTimes(0);
});

test('MonthCycler can accomodate arbitrary date range', () => {
  let onChange = jest.fn();
  let value = { date: new Date(), onChange };
  let { container, getByText } = render(
    <DateContext.Provider value={value}>
      <MonthCycler />
      <Calendar />
    </DateContext.Provider>
  );
  // debug();
  let i = 1000;
  while (i--) {
    fireEvent.click(getByText('←'));
  }
  expect(onChange).toHaveBeenCalledTimes(1000);
});

test('MonthCycler accomodates className', () => {
  let className = 'c';
  let { container } = renderWithContext(MonthCycler, {}, { className });
  expect(container.querySelector('div').hasAttribute('class')).toBe(true);
});

test('MonthCycler throws if rendered without context', () => {
  expect(() => render(<MonthCycler />)).toThrow();
});

test('Calendar throws if rendered without context', () => {
  expect(() => render(<Calendar />)).toThrow();
});

test('Calendar accomodates className', () => {
  let className = 'c';
  let { container, debug } = renderWithContext(Calendar, {}, { className });
  expect(container.querySelector('table').hasAttribute('class')).toBe(true);
});

test('Calendar can render day names', () => {
  let { getByText } = renderWithContext(Calendar, {}, { weekdays: true });
  dayNames.forEach(d => expect(getByText(d)).not.toBeUndefined());
});

test('Calendar renders array of weekdays', () => {
  let weekdays = ['S', 'M', 'T'];
  let { container } = renderWithContext(Calendar, {}, { weekdays });
  expect(container.querySelector('tr').textContent).toBe(weekdays.join(''));
});

test('Calendar passes selected date to onChange', () => {
  let onChange = jest.fn();
  let { container, getByText } = renderWithContext(Calendar, { onChange });
  expect(container.querySelector('td')).not.toBeUndefined();
  fireEvent.click(getByText('1'));
  expect(onChange).toHaveBeenCalled();
});
