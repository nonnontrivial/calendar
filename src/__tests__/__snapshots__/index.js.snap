// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar handles obscure dates 1`] = `
<table>
  <tbody>
    <tr>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: default;"
      />
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: default;"
      />
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: default;"
      />
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: default;"
      />
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: default;"
      />
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: default;"
      />
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        1
      </td>
    </tr>
    <tr>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        2
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        3
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        4
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        5
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        6
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        7
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        8
      </td>
    </tr>
    <tr>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        9
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        10
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        11
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        12
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        13
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        14
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        15
      </td>
    </tr>
    <tr>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        16
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        17
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        18
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer; box-shadow: 0 4px 6px 2px rgba(0, 0, 0, .10);"
      >
        19
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        20
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        21
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        22
      </td>
    </tr>
    <tr>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        23
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        24
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        25
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        26
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        27
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        28
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        29
      </td>
    </tr>
    <tr>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        30
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: pointer;"
      >
        31
      </td>
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: default;"
      />
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: default;"
      />
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: default;"
      />
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: default;"
      />
      <td
        style="text-align: center; padding: 0.421rem; border-radius: 0.221rem; cursor: default;"
      />
    </tr>
  </tbody>
</table>
`;

exports[`MonthCycler uses locale passed to context 1`] = `
<p>
  Dec
   
  2018
</p>
`;

exports[`defaults to current date if none provided 1`] = `
<p>
  Dec
   
  2018
</p>
`;

exports[`displays date as string 1`] = `
<p>
  Dec
   
  2018
</p>
`;
